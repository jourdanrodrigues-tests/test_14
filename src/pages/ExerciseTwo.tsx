import React, { useMemo } from 'react';

import { Card, Typography } from '@material-tailwind/react';

import Page from '@/components/Page.tsx';
import { useExtractions } from '@/exercise2Api/queries.ts';
import { useNavigator } from '@/routes.ts';

export default function ExerciseTwo() {
  const navigator = useNavigator();
  const extractions = useExtractions();
  const { completed, pending, failed } = useMemo(() => {
    const output = { completed: 0, pending: 0, failed: 0 };
    for (const extraction of extractions) {
      output[extraction.status]++;
    }
    return output;
  }, [extractions]);

  return (
    <Page
      title={`Exercise #2 - Fetched ${extractions.length} extractions`}
      className="gap-10 justify-start mt-10"
    >
      <button onClick={navigator.home}>Go back</button>
      <Typography>
        Completed: {completed} | Pending: {pending} | Failed: {failed}
      </Typography>
      <Card className="h-60 overflow-y-auto">
        <table className="table-auto relative w-full border">
          <thead>
            <tr className="bg-blue-gray-50 sticky top-0">
              <Header>ID</Header>
              <Header>Configuration</Header>
              <Header>Status</Header>
            </tr>
          </thead>
          <tbody className="divide-y">
            {extractions.map((extraction) => (
              <tr key={extraction.id}>
                <Cell>{extraction.id}</Cell>
                <Cell>{extraction.configuration}</Cell>
                <Cell>{extraction.status}</Cell>
              </tr>
            ))}
          </tbody>
        </table>
      </Card>
    </Page>
  );
}

function Header({ children }: { children: React.ReactNode }) {
  return <th className="px-6 py-3">{children}</th>;
}

function Cell({ children }: { children: React.ReactNode }) {
  return (
    <td className="py-3 px-4">
      <Typography>{children}</Typography>
    </td>
  );
}
