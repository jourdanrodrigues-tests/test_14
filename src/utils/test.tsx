import React from 'react';

import type { UseQueryResult } from '@tanstack/react-query';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render as rtlRender, waitFor } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import { expect } from 'bun:test';
// @ts-expect-error - No need to have "history" as an explicit dependency
import type { InitialEntry } from 'history';
import pretty from 'pretty';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

type TQueryHooks<
  TData = unknown,
  TError = unknown,
  QueryResult = UseQueryResult<TData, TError>,
> = Array<{
  useHook(): QueryResult;
  validate(query: QueryResult): boolean;
}>;

type RenderOptions = {
  queryHooks?: TQueryHooks;
  route?: [InitialEntry, string];
};

export const testQueryClient = new QueryClient();
const queryClientWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

export async function render(
  node: React.ReactElement,
  options?: RenderOptions
) {
  const hookPromises = (options?.queryHooks || []).map(
    ({ useHook, validate }) => {
      const { result } = renderHook(useHook, { wrapper: queryClientWrapper });
      return waitFor(() => validate(result.current));
    }
  );
  if (hookPromises.length > 0) {
    await Promise.all(hookPromises);
  }

  return rtlRender(node, {
    wrapper: ({ children }) => wrapNode(children, options),
  });
}

function wrapNode(
  node: React.ReactNode,
  options?: RenderOptions
): React.ReactNode {
  const { route } = options || {};
  if (route) {
    const [entry, path] = route;
    node = (
      <MemoryRouter initialEntries={[entry]}>
        <Routes>
          <Route path={path} element={node} />
        </Routes>
      </MemoryRouter>
    );
  }

  return queryClientWrapper({ children: node });
}

export function expectSnapshot(element: HTMLElement | null) {
  // TODO: Use @testing-library/jest-dom instead of @happy-dom
  //  The JestDOM does some HTML pretty-printing that is better for diffs
  return expect(element && pretty(element.innerHTML));
}
