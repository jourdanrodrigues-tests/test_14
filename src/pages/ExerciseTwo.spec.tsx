import { afterEach, describe, it, mock } from 'bun:test';

import type { Extraction } from '@/exercise2Api';
import ExerciseTwo from '@/pages/ExerciseTwo.tsx';
import { expectSnapshot, render } from '@/utils/test.tsx';

describe('ExerciseTwo Component', () => {
  afterEach(() => {
    mock.restore();
  });

  it('should match the snapshot', async () => {
    const now = Date.now();
    const extractions: Extraction[] = [
      {
        id: 'ID_1',
        configuration: 'Configuration 1',
        created: now.toLocaleString(),
        status: 'completed',
      },
      {
        id: 'ID_2',
        configuration: 'Configuration 2',
        created: now.toLocaleString(),
        status: 'pending',
      },
      {
        id: 'ID_3',
        configuration: 'Configuration 3',
        created: now.toLocaleString(),
        status: 'failed',
      },
    ];

    mock.module('@/exercise2Api/hooks', () => ({
      useExtractions: () => extractions,
    }));

    const result = await render(<ExerciseTwo />, {
      route: ['/exercise2', '/exercise2'],
    });

    expectSnapshot(result.container).toMatchSnapshot();
  });
});
