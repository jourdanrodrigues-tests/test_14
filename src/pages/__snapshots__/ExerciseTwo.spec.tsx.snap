// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseTwo Component should match the snapshot 1`] = `
"<div class="max-h-screen w-screen flex flex-col items-center gap-10 justify-start mt-10">
  <h1 class="block antialiased tracking-normal font-sans text-5xl font-semibold leading-tight text-inherit">Exercise #2 - Fetched 3 extractions</h1><button>Go back</button>
  <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Completed: 1 | Pending: 1 | Failed: 1</p>
  <div class="relative flex flex-col bg-clip-border rounded-xl bg-white text-gray-700 shadow-md h-60 overflow-y-auto">
    <table class="table-auto relative w-full border">
      <thead>
        <tr class="bg-blue-gray-50 dark:bg-blue-gray-800 sticky top-0">
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">ID</p>
            </div>
          </th>
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Configuration</p>
            </div>
          </th>
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Date</p>
              <div class="flex flex-col">
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
              </div>
            </div>
          </th>
          <th class="px-6 py-3">
            <div class="flex gap-4">
              <p class="block antialiased font-sans text-base font-light leading-relaxed text-inherit">Status</p>
              <div class="flex flex-col">
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▲</p>
                <p class="block antialiased font-sans font-light text-inherit text-xs leading-none cursor-pointer select-none">▼</p>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody class="divide-y">
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">
            <p class="block antialiased font-light text-inherit font-mono text-sm">ID_1</p>
          </td>
          <td class="py-3 px-4">Configuration 1</td>
          <td class="py-3 px-4">Invalid Date</td>
          <td class="py-3 px-4">Completed</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">
            <p class="block antialiased font-light text-inherit font-mono text-sm">ID_2</p>
          </td>
          <td class="py-3 px-4">Configuration 2</td>
          <td class="py-3 px-4">Invalid Date</td>
          <td class="py-3 px-4">Pending</td>
        </tr>
        <tr class="even:bg-blue-gray-50/50">
          <td class="py-3 px-4">
            <p class="block antialiased font-light text-inherit font-mono text-sm">ID_3</p>
          </td>
          <td class="py-3 px-4">Configuration 3</td>
          <td class="py-3 px-4">Invalid Date</td>
          <td class="py-3 px-4">Failed</td>
        </tr>
      </tbody>
    </table>
  </div>
</div>"
`;
